Use help_formatter to consistently format help output

tautschnig opened the pull request:
We previously did manual line breaks that either suited the source code
line length or approximated the output line width. help_formatter, however,
was specifically built to take care of this (and just wasn't widely
used). We can then also happily have clang-format take care of source
code formatting without impacting the resulting display.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
Use help_entry to consistently format help output

We previously did manual line breaks that either suited the source code
line length or approximated the output line width. help_entry, however,
was specifically built to take care of this (and just wasn't widely
used). We can then also happily have clang-format take care of source
code formatting without impacting the resulting display.

martin-cs commented on jbmc/src/janalyzer/janalyzer_parse_options.cpp:
```diff
@@ -727,77 +727,86 @@ void janalyzer_parse_optionst::help()
-    " --dependence-graph           data and control dependencies between instructions\n" // NOLINT(*)
-    "\n"
+    << help_entry("--constants", "constant domain")
+    << help_entry("--intervals, --show-intervals", "interval domain")
```
Will revert https://github.com/diffblue/cbmc/pull/6932

tautschnig replied:
Yes, I'm hoping to get #6932 approved first (via @peterschrammel) and will then rebase this one on top to get rid of `--show-intervals` in here.

martin-cs commented on src/analyses/variable-sensitivity/variable_sensitivity_domain.h:
```diff
@@ -80,26 +80,28 @@
-  "(vsd-liveness)"                                                             \
-                                                                               \
-    // clang-format off
+  "(vsd-liveness)"
```
Are we still doing the "put a \ at the end of even the last line so that new additions are just 1-line diffs" thing?

tautschnig replied:
Might be useful, but clang-format takes them away. So I'd have to (re-)add `clang-format off/on`, which makes formatting more cumbersome. I'm not sure which is best, happy to take guidance!

martin-cs commented on src/analyses/variable-sensitivity/variable_sensitivity_domain.h:
```diff
@@ -80,26 +80,28 @@
-  " --vsd-liveness               track variable liveness\n"                    \
-
-// cland-format on
+  help_entry(                                                                  \
```
I feel a bit weird about having the `<<` operators for all but the first.  I guess we have to `out << HELP_VSD` which is grammatical but ... it feels a bit weird.

tautschnig replied:
I'm happy to move the `<<` over into the macros, but I don't know whether it actually makes for less weirdness. I tripped over this myself when coming up with this patch...

martin-cs commented on src/goto-analyzer/goto_analyzer_parse_options.cpp:
```diff
@@ -695,107 +695,123 @@ void goto_analyzer_parse_optionst::help()
-    "\n"
+    << help_entry("--constants", "a constant for each variable if possible")
+    << help_entry(
+      "--intervals, --show-intervals", "an interval for each variable")
```
Also a potential revert here.

tautschnig replied:
Ack.

martin-cs reviewed (approved):
Good stuff but be aware of the potential revert of #6932

NlightNFotis reviewed (approved):
Hi Michael.

This looks tedious, but it's a very important step to establish some uniformity and a standard around the help options.

Thanks for taking the time to sort this out.

tautschnig commented:
Pinging @chris-ryder as I believe I'll strictly need your approval here.

feliperodri reviewed (approved):
Changes in `contracts` LGTM.

remi-delmas-3000 commented on src/util/config.h:
```diff
@@ -79,39 +97,41 @@ class symbol_tablet;
-  "(object-bits):"                                                             \
+  "(gcc)"
+
+#define HELP_CONFIG_PLATFORM                                                   \
```
Since help_entry is a function could we instead write these as c++ functions returning `std::string` and avoid the multi-line macros ?

remi-delmas-3000 reviewed (commented):
Are the multi-line macros that just expand chais of calls to `help_entry` and concatenation using `<<` (see `#define HELP_CONFIG_LIBRARY` for instance) really needed or could we just use C++ functions returning `std::string` everywhere ?

remi-delmas-3000 reviewed (approved):
Are the multi-line macros that just expand chais of calls to `help_entry` and concatenation using `<<` (see `#define HELP_CONFIG_LIBRARY` for instance) really needed or could we just use C++ functions returning `std::string` everywhere ?

tautschnig commented:
> Are the multi-line macros that just expand chais of calls to `help_entry` and concatenation using `<<` (see `#define HELP_CONFIG_LIBRARY` for instance) really needed or could we just use C++ functions returning `std::string` everywhere ?

I'll work on this.

tautschnig commented:
I should likely rebase/rework this to use #7141.

tautschnig committed:
Use help_formatter

First round of changes was done automatically:
```
sed -i 's/help_entry("\(--[^ ]*\)", "\(.*\)")/" {y\1} \\t \2\\n"/' \
  src/*/*.h src/*/*.cpp src/analyses/variable-sensitivity/*.h \
  jbmc/src/*/*.h jbmc/src/*/*.cpp
```
All further replacements were done manually.

tautschnig committed:
Remove help_entry, which is now unused

All uses of `help_entry` have been converted to uses of
`help_formatter`.

tautschnig committed:
Add unit test of help_formatter

Modelled after help_entry's unit test.

tautschnig commented:
To anyone reviewing this: on this occasion, reviewing the overall diff will be more effective than doing individual commits (I'm mainly keeping the "Use help_entry" commit for sentimental reasons).
